File: 22 Generate Parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        M = n;
        f("", 0);
        return re;
    }
private:
    void f(string pre, int l){
        if(l == M){
            re.push_back(pre + string(2 * M - pre.size(), ')'));
            return;
        }
        for(;pre.size() <= 2 * l;pre += ')'){
            f(pre + '(', l + 1);
        }
    }
    int M;
    vector<string> re; 
};

File: 22 Generate Parentheses.py
class Solution:
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        re = []
        def f(pre, l):
            if l == n:
                re.append(pre + ')' * (2 * n - len(pre)))
                return 0
            while len(pre) <= 2 * l:
                f(pre + '(', l + 1)
                pre += ')'
        f("", 0)
        return re